// @sap/ams/lib/express/middleware.js
"use strict";

const PolicyDecisionPoint = require("../policyDecisionPoint");
const ErrType = require("../errors/errorTypes");
const Attributes = require("../attributes");
const AttributeName = require("../attributeName");

const STARTUP_CHECK_TIMEOUT = 30000; // keep in sync with Java AMS lib

let pdp = null;

function middlewareHelper(attributes, failMsg, addPrincipal) {
	const pdp = getPdp(); // trigger bundle loading in background on initialization

	return async function(req, res, next) {
		await pdp.startupCheck(STARTUP_CHECK_TIMEOUT);

		if (addPrincipal) {
			if (!req.tokenInfo) {
				next(new ErrType.AuthenticationError("User token not found.", 401));
			}
			attributes.setTokenInfo(req.tokenInfo);
		}
		pdp.allow(attributes).then(
			(allowed) => {
				if (allowed) {
					next();
				} else {
					next(new ErrType.AuthorizationError(failMsg, 403));
				}
			},
			(err) => {
				next(new ErrType.RethrowError(`Allow request failed: ${err}`, 500, err));
			});
	};
}

function getPdp() {
	pdp ??= new PolicyDecisionPoint();

	return pdp;
};

/**
 * Return the PolicyDecisionPoint instance which is useful for
 * for accessing the PolicyDecisionPoint events
 */
module.exports.getPdp = getPdp;

/**
 * Checks if a user is allowed regarding the given attributes
 * @param {Attributes} attributes
 * @param {bool} addPrincipal if true adds token user info (principal2policy) to attributes
 */
module.exports.isAllowed = function(attributes, addPrincipal) {
	const failMsg = "User is not allowed";
	return middlewareHelper(attributes, failMsg, addPrincipal);
};

/**
 * This check can only be applied for very trivial policies e.g.:
 * POLICY myPolicy { GRANT action ON *; }.
 * @param {string} action
 * returns next() if the authenticated user has the permission to perform the given action. 
 */
module.exports.hasAuthority = function(action) {
	const attributes = new Attributes()
		.setAction(action);
	const failMsg = `User does not have authority for action: ${action}`;
	return middlewareHelper(attributes, failMsg, true);
};

/**
 * @param {string} action
 * @param {string} resource
 * returns next() if the authenticated user has in principal the permission 
 * to perform a given action on a given resource. 
 * It's also allowed to use a '*' to be less restrictive.
 */
module.exports.hasBaseAuthority = function(action, resource) {
	const ignores = [["$app"]];
	if (action == "*")
		ignores.push(["$dcl", "action"]);
	if (resource == "*")
		ignores.push(["$dcl", "resource"]);
	const attributes = new Attributes()
		.setAction(action)
		.setResource(resource)
		.setIgnores(ignores);
	const failMsg = `User does not have authority for action: ${action} and resource: ${resource}`;
	return middlewareHelper(attributes, failMsg, true);
};

/**
 * @param {string} action
 * @param {object} app Optionally you can restrict to attributes by providing 
 * one or more attributes in the parameter app e.g. {"CountryCode": "GB", "salesOrderId": 233}.
 * returns next() if the user is authorized to perform a dedicated action. 
 * The resource gets ignored. 
 */
module.exports.forAction = function(action, app) {
	const attributes = new Attributes()
		.setAction(action)
		.setApp(app)
		.addIgnores(AttributeName.common.DCL_RESOURCE);
	const failMsg = extendFailMessage(`User does not have authority for action: ${action}`, app);
	return middlewareHelper(attributes, failMsg, true);
};

/**
 * @param {string} resource
 * @param {object} app
 * returns next() if the user is authorized to access the given resource. 
 * The action gets ignored. 
 * Optionally you can restrict to attributes by providing additional key, value pairs via the parameter app
 */
module.exports.forResource = function(resource, app) {
	const attributes = new Attributes()
		.setResource(resource)
		.setApp(app)
		.addIgnores(AttributeName.common.DCL_ACTION);
	const failMsg = extendFailMessage(`User does not have authority for resource: ${resource}`, app);
	return middlewareHelper(attributes, failMsg, true);
};

/**
 * @param {string} resource
 * @param {string} action
 * @param {object} app
 * returns true if the user is authorized to perform a dedicated action on the given resource.
 * Optionally you can restrict to attributes by providing additional key, value pairs via the parameter app
 */
module.exports.forResourceAction = function(resource, action, app) {
	const attributes = new Attributes()
		.setResource(resource)
		.setAction(action)
		.setApp(app);
	const failMsg = extendFailMessage(`User does not have authority for action: ${action} on resource: ${resource}`, app);
	return middlewareHelper(attributes, failMsg, true);
};

function extendFailMessage(msg, app) {
	if (app) {
		msg += ` where: ${JSON.stringify(app, null, 2)}`;
	}
	return msg;
}
