// @sap/ams/src/xcode/dcn-engine/build/commonjs/legacy/tranformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.emulateOPA = emulateOPA;
exports.transformDCNtoOPA = transformDCNtoOPA;
var _and = _interopRequireDefault(require("../core/condition/and.js"));
var _equal = _interopRequireDefault(require("../core/condition/equal.js"));
var _false = _interopRequireDefault(require("../core/condition/false.js"));
var _optimize = _interopRequireDefault(require("../optimize.js"));
var _dcl = require("../util/dcl.js");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 *  
 * @param {Decision} decision 
 */
function emulateOPA(decision) {
  const optimized = preemptiveOptimization(decision.__internal__condition);
  return transformDCNtoOPA(optimized.toDcn());
}
function preemptiveOptimization(condition) {
  // return Optimizer.optimize(condition);
  if (condition instanceof _and.default) {
    if (condition.conditions[0] instanceof _equal.default && condition.conditions[1] instanceof _equal.default) {
      if (condition.conditions[0].values[0].reference === condition.conditions[1].values[0].reference) {
        if (condition.conditions[0].values[1].constant !== condition.conditions[1].values[1].constant) {
          return new _false.default();
        }
      }
    }
  }
  return condition;
}
function legacyArgumentsOrder(a, b) {
  if (typeof a === "object" && typeof b !== "object") {
    return -1;
  } else if (typeof a !== "object" && typeof b === "object") {
    return 1;
  }
  if (a.args?.[0] < b.args?.[0]) {
    return -1;
  } else if (a.args?.[0] > b.args?.[0]) {
    return 1;
  } else if (a.args?.[1] < b.args?.[1]) {
    return -1;
  } else if (a.args?.[1] > b.args?.[1]) {
    return 1;
  } else if (a.args?.[0]?.ref < b.args?.[0]?.ref) {
    return -1;
  } else if (a.args?.[0]?.ref > b.args?.[0]?.ref) {
    return 1;
  } else if (a.args?.[1]?.ref < b.args?.[1]?.ref) {
    return -1;
  } else if (a.args?.[1]?.ref > b.args?.[1]?.ref) {
    return 1;
  }
  return 0;
}
function removeDuplicates(value, index, array) {
  return index === 0 || JSON.stringify(value) !== JSON.stringify(array[index - 1]);
}
function transformDCNtoOPA(dcn) {
  if (!Array.isArray(dcn.call)) {
    return dcn;
  }
  if (dcn.call.length != 1) {
    return dcn;
  }
  let args;
  switch (dcn.call[0]) {
    case "eq":
      return {
        call: "eq",
        args: dcn.args.sort(legacyArgumentsOrder)
      };
    case "ne":
      return {
        call: "ne",
        args: dcn.args.sort(legacyArgumentsOrder)
      };
    case "or":
      // Flatten nested ors
      args = dcn.args.map(transformDCNtoOPA).reduce((acc, arg) => {
        if (arg.call === "or") {
          return acc.concat(arg.args);
        }
        return acc.concat([arg]);
      }, []).sort(legacyArgumentsOrder).filter(removeDuplicates);
      return args.length === 1 ? args[0] : {
        call: "or",
        args: args
      };
    case "and":
      // Flatten nested ands
      args = dcn.args.map(transformDCNtoOPA).reduce((acc, arg) => {
        if (arg.call === "and") {
          return acc.concat(arg.args);
        }
        return acc.concat([arg]);
      }, []).sort(legacyArgumentsOrder).filter(removeDuplicates);
      return args.length === 1 ? args[0] : {
        call: "and",
        args: args
      };
    case "in":
      if (!Array.isArray(dcn.args[1])) {
        return dcn;
      }
      args = dcn.args[1].map(element => {
        return {
          call: 'eq',
          args: [dcn.args[0], element]
        };
      });
      return args.length === 1 ? args[0] : {
        call: "or",
        args: args
      };
    case "not_in":
      if (!Array.isArray(dcn.args[1])) {
        return dcn;
      }
      args = dcn.args[1].map(element => {
        return {
          call: 'ne',
          args: [dcn.args[0], element]
        };
      });
      return args.length === 1 ? args[0] : {
        call: "and",
        args: args
      };
    default:
      dcn.call = (0, _dcl.stringifyReference)(dcn.call);
      return dcn;
  }
}
