// @sap/ams/lib/errors/errorTypes.js
"use strict";

const BaseError = require("./baseError");

/**
 * E.g. User token is missing or invalid
 */
class AuthenticationError extends BaseError {  
	constructor (message, code) { super(message, code); }
}

/**
 * E.g. User is not allowed to read a resource
 */
class AuthorizationError extends BaseError {  
	constructor (message, code) { super(message, code); }
}

/**
 * E.g. Opa url is not reachable
 */
class OpaConnError extends BaseError {  
	constructor (message, code) { super(message, code); }
}

class DCLError extends BaseError {  
	constructor (message, code) { super(message, code); }
}

class AttributesError extends BaseError {  
	constructor (message, code) { super(message, code); }
}

/**
 * DCL PackageName may not start with number, be empty or end with '.'
 */
class DCLPackageNameError extends BaseError {  
	constructor (message, code) { super(message, code); }
}

class RethrowError extends BaseError {
	constructor(message, code, error){
		super(message, code);
		
		if (!error) throw new Error("RethrowError requires a message and error");
		this.original_error = error;
		this.stack_before_rethrow = this.stack;
		this.stack = this.stack + "\n" + error.stack;
	}
}

class EngineStartupCheckError extends BaseError {
	constructor (message, code) { super(message, code); }
}

class CredentialsConfigurationError extends BaseError {}
  
module.exports = { 
	AuthenticationError, 
	AuthorizationError,
	CredentialsConfigurationError,
	OpaConnError,
	DCLError,
	DCLPackageNameError,
	RethrowError,
	AttributesError,
	EngineStartupCheckError
};
