// @sap/ams/lib/runtime/AmsInstance.js
const uploadBaseDCL = require('./uploader');
const writeOPAConfig = require('./opaconfigurator');
const downloadBundle = require('./downloader');
const extractBundle = require('./extractor');

function validateIdentityCredentials(iasCredentials) {
    if (!iasCredentials) {
      throw new Error("Missing identity credentials.");
    }
  
    const requiredAttributes = [
      "certificate",
      "key",
      "authorization_instance_id",
      "url"
    ];
  
    for(const a of requiredAttributes) {
      if (!iasCredentials[a]) {
        throw new Error(`Mandatory attibute "${a}" is missing from identity credentials.`);
      }
    }

    return iasCredentials;
}

class AmsInstance {
    #credentials;

    constructor(identity_credentials) {
        this.#credentials = validateIdentityCredentials(identity_credentials);
    }

    async uploadBaseDCL(ams_root) {
        return uploadBaseDCL(this.#credentials, ams_root);
    }

    async writeOPAConfig(opa_root) {
        return writeOPAConfig(this.#credentials, opa_root);
    }

    async downloadBundle() {
        return downloadBundle(this.#credentials);
    }

    async extractBundle(bundle, path_root) {
        return extractBundle(bundle, path_root);
    }
}

module.exports = AmsInstance;
