// @sap/ams/lib/runtime/common.js
const https = require("https");
const stream = require("stream");
const fs = require('fs');

const { createHash } = require("crypto");
const readdir = fs.readdir;

let USER_AGENT = "ams.nodejs.lib";

const AMS_INSTANCE_HEADER = "X-Ams-Instance-Id";


class InMemoryStream extends stream.Writable {
  #buffer;
  constructor() {
      super();
      this.#buffer = Buffer.alloc(0);
  }

  _write(chunk, enc, next) {
      this.#buffer = Buffer.concat([this.#buffer, chunk]); 
      next();
  }

  getBuffer() {
      return this.#buffer;
  }

  getString() {
      return this.#buffer.toString('utf-8');
  }

  getHash() {
    return createHash("sha256").update(this.#buffer).digest("hex");
  }  
}

module.exports = {
  setUserAgent(agent) {
    USER_AGENT = agent;
  },
  createHTTPSAgent(options) {
    return new https.Agent({
        cert: options.certificate,
        key: options.key,
    });
  },
  addDefaultHeaders(headers, identity) {
    headers[AMS_INSTANCE_HEADER] = identity.authorization_instance_id;
    headers["User-Agent"] = USER_AGENT;
  },
  async filterDCLDirectory(ams_root, cb) {    
    fs.readdir = function (path, cb, val3) {
      if(path.startsWith(ams_root)) {
        return readdir(path, async (err, files) => {
          if(err) {
            return cb(err);
          }
          const filteredFiles = [];
          for(let i=0;i<files.length;++i) {
            const f = files[i];
            const stat = await fs.promises.stat(path + "/" + f);
            if(stat.isFile()) {
              if(f.endsWith(".dcl")) {
                filteredFiles.push(f);
              }
            } else {
              filteredFiles.push(f);
            }
          }

          cb(null, filteredFiles);
        });
      } else {
        return readdir(path, cb, val3);
      }
    };
    await cb();
    fs.readdir = readdir;
  },
  InMemoryStream
}
