// @sap/ams/lib/runtime/opaconfigurator.js
const fs = require("fs/promises");
const path = require("path");

const yaml = require("json-to-pretty-yaml");

const OPA_SERVICE_KEY = "bundle_storage";
const CERT_FILE_NAME = "certificate.pem"
const KEY_FILE_NAME = "key.pem"
const CONFIG_FILE_NAME = "config.yml"

function getConfigJSON(identity, cert, key) {
  const obj = {
    bundles: {},
    services: {},
    plugins: {
      dcl: true,
    },
    status: {
      plugin: "dcl",
    },
  };

  //Add the bundles entry
  obj.bundles[identity.authorization_instance_id] = {
    service: OPA_SERVICE_KEY,
    resource: identity.authorization_instance_id + ".tar.gz",
    polling: {
      min_delay_seconds: 10,
      max_delay_seconds: 20
    },
  };

  //add the service Entry
  obj.services[OPA_SERVICE_KEY] = {
    url: identity.authorization_bundle_url,
    headers: {
      AMS_INSTANCE_HEADER: identity.authorization_instance_id,
    },
    credentials: {
      client_tls: {
        cert: cert,
        private_key: key,
      },
    },
  };

  return obj;
}

function getConfig(identity, opa_root) {
    const cert = path.join(opa_root, CERT_FILE_NAME);
    const key = path.join(opa_root, KEY_FILE_NAME);

    return {
      serviceKey: OPA_SERVICE_KEY,
      opa: getConfigJSON(identity, cert, key),
      paths: {
        cert: cert,
        key: key,
        config: path.join(opa_root, CONFIG_FILE_NAME),
      },
      identity: identity,
      opa_root: opa_root,
    };
}

module.exports = async function(credentials, opa_root) {
    const config_json = getConfig(credentials, opa_root ? opa_root : "./");

    //create yaml string
    const ret = yaml.stringify(config_json.opa);

    if(opa_root) {
      const key_prom = fs.writeFile(config_json.paths.key, credentials.key);
      const cert_prom = fs.writeFile(config_json.paths.cert, credentials.certificate);
      const config_prom = await fs.writeFile(config_json.paths.config, ret);
      await Promise.all([key_prom, cert_prom, config_prom]);
    }

    return { yaml: ret, config: config_json };
}
