// @sap/ams/lib/runtime/uploader.js

const stream = require("stream");
const compressing = require("compressing");
const {addDefaultHeaders, createHTTPSAgent, filterDCLDirectory, InMemoryStream} = require('./common');
const httpClient = require("./httpClient");

async function prepareUploadDCL(identity, ams_root) {
  //try to compress the whole directory into memory buffer  
  const buf = new InMemoryStream();

  await filterDCLDirectory(ams_root, async ()=> {    
    return compressing.tgz.compressDir(ams_root, buf, { ignoreBase: true });
  });

  //return a config object suiteable for all http clients
  const ret = {
    url: `${identity.url}/sap/ams/v1/ams-instances/${identity.authorization_instance_id}/dcl-upload`,
    method: "POST",
    host: identity.url,
    headers: {
      "Content-Type": "application/gzip"
    },
    agent: createHTTPSAgent(identity),
    body: buf.getBuffer(),
  };

  addDefaultHeaders(ret.headers, identity);
  return ret;
}

module.exports = async function(credentials, ams_root) {
  const config = await prepareUploadDCL(credentials, ams_root);
  return httpClient.uploadDCL(config);
}
