// @sap/ams/src/xcode/dcn-engine/build/commonjs/core/condition/equal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _comparisonoperator = _interopRequireDefault(require("./comparisonoperator.js"));
var _false = _interopRequireDefault(require("./false.js"));
var _function = _interopRequireDefault(require("./function.js"));
var _not = _interopRequireDefault(require("./not.js"));
var _true = _interopRequireDefault(require("./true.js"));
var _value = _interopRequireDefault(require("./value.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class Equal extends _comparisonoperator.default {
  static fromDcn(dcnArguments = {}, schema) {
    const v1 = _value.default.fromDcn(dcnArguments[0], schema);
    const v2 = _value.default.fromDcn(dcnArguments[1], schema);
    return new Equal(v1, v2);
  }
  evaluate(input = {}) {
    const v1 = this.values[0].resolve(input);
    const v2 = this.values[1].resolve(input);
    if (v1 instanceof _function.default) {
      if (v2.constant === true) {
        return v1.condition;
      }
    }
    if (v2 instanceof _function.default) {
      if (v1.constant === true) {
        return v2.condition;
      }
    }
    if (v1.constant === undefined || v2.constant === undefined) {
      return new Equal(v1, v2);
    } else {
      if (v1.constant === v2.constant) {
        return new _true.default();
      } else {
        return new _false.default();
      }
    }
  }
  toDcn() {
    return {
      call: ["eq"],
      args: this.values.map(v => v.toDcn())
    };
  }
}
exports.default = Equal;
