// @sap/ams/src/xcode/dcn-engine/build/commonjs/core/condition/is_null.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _comparisonoperator = _interopRequireDefault(require("./comparisonoperator.js"));
var _false = _interopRequireDefault(require("./false.js"));
var _true = _interopRequireDefault(require("./true.js"));
var _value = _interopRequireDefault(require("./value.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class IsNull extends _comparisonoperator.default {
  static fromDcn(dcnArguments = {}, schema) {
    return new IsNull(_value.default.fromDcn(dcnArguments[0], schema));
  }
  evaluate(input = {}) {
    const v = this.values[0];
    // remark null != NULL
    if (v.constant || v.constant === null) {
      return new _false.default();
    }
    if (Object.prototype.hasOwnProperty.call(input, v.reference)) {
      if (input[v.reference] === undefined) {
        return new _true.default();
      } else {
        return new _false.default();
      }
    }
    return this;
  }
  toDcn() {
    return {
      call: ["is_null"],
      args: [this.values[0].toDcn()]
    };
  }
}
exports.default = IsNull;
