// @sap/ams/src/xcode/dcn-engine/build/commonjs/core/condition/not.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _condition = _interopRequireDefault(require("./condition.js"));
var _false = _interopRequireDefault(require("./false.js"));
var _true = _interopRequireDefault(require("./true.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class Not extends _condition.default {
  /** 
   * @type {Condition}
   */
  condition = undefined;
  /**
   * @param {Condition}
   */
  constructor(condition) {
    super();
    this.condition = condition;
    this.variables = condition.variables;
  }
  evaluate(input = {}) {
    const evaluated = this.condition.evaluate(input);
    if (evaluated.isTrue()) {
      return new _false.default();
    }
    if (evaluated.isFalse()) {
      return new _true.default();
    }
    return new Not(evaluated);
  }
  toDcn() {
    return {
      call: ["not"],
      args: [this.condition.toDcn()]
    };
  }
  static create(condition) {
    return new Not(condition);
  }
}
exports.default = Not;
