// @sap/ams/src/xcode/dcn-engine/build/commonjs/core/condition/not_in.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _comparisonoperator = _interopRequireDefault(require("./comparisonoperator.js"));
var _false = _interopRequireDefault(require("./false.js"));
var _true = _interopRequireDefault(require("./true.js"));
var _value = _interopRequireDefault(require("./value.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class NotIn extends _comparisonoperator.default {
  static fromDcn(dcnArguments = {}, schema) {
    return new NotIn(_value.default.fromDcn(dcnArguments[0], schema), _value.default.fromDcn(dcnArguments[1], schema));
  }
  evaluate(input = {}) {
    const v1 = this.values[0].resolve(input);
    const v2 = this.values[1].resolve(input);
    if (v2.constant === undefined) {
      return new NotIn(v1, v2);
    }
    if (v2.constant.length === 0) {
      return new _true.default();
    }
    if (v1.constant === undefined) {
      return new NotIn(v1, v2);
    } else {
      if (v2.constant.indexOf(v1.constant) !== -1) {
        return new _false.default();
      } else {
        return new _true.default();
      }
    }
  }
  toDcn() {
    return {
      call: ["not_in"],
      args: this.values.map(v => v.toDcn())
    };
  }
}
exports.default = NotIn;
