// @sap/ams/src/xcode/dcn-engine/build/commonjs/core/condition/not_between.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _comparisonoperator = _interopRequireDefault(require("./comparisonoperator.js"));
var _false = _interopRequireDefault(require("./false.js"));
var _true = _interopRequireDefault(require("./true.js"));
var _value = _interopRequireDefault(require("./value.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class NotBetween extends _comparisonoperator.default {
  static fromDcn(dcnArguments = {}, schema) {
    return new NotBetween(_value.default.fromDcn(dcnArguments[0], schema), _value.default.fromDcn(dcnArguments[1], schema), _value.default.fromDcn(dcnArguments[2], schema));
  }
  evaluate(input = {}) {
    const v1 = this.values[0].resolve(input);
    const v2 = this.values[1].resolve(input);
    const v3 = this.values[2].resolve(input);
    if (v2.constant === undefined || v3.constant === undefined) {
      return new NotBetween(v1, v2, v3);
    }
    if (v2.constant > v3.constant) {
      return new _true.default();
    }
    if (v1.constant === undefined) {
      return new NotBetween(v1, v2, v3);
    }
    if (v1.constant < v2.constant || v1.constant > v3.constant) {
      return new _true.default();
    } else {
      return new _false.default();
    }
  }
  toDcn() {
    return {
      call: ["not_between"],
      args: this.values.map(v => v.toDcn())
    };
  }
}
exports.default = NotBetween;
