// @sap/ams/src/xcode/dcn-engine/build/commonjs/core/condition/or.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _condition = _interopRequireDefault(require("./condition.js"));
var _false = _interopRequireDefault(require("./false.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class Or extends _condition.default {
  /** 
   * @type {Condition[]}
   */
  conditions = [];
  /**
   * @param {Condition[]}
   */
  constructor(conditions) {
    super();
    this.conditions = conditions;
    this.variables = conditions.reduce((p, c) => p.concat(c.variables), []);
  }
  evaluate(input = {}) {
    const conditions = [];
    for (const c of this.conditions) {
      const evaluated = c.evaluate(input);
      if (evaluated.isTrue()) {
        return evaluated;
      }
      if (!evaluated.isFalse()) {
        conditions.push(evaluated);
      }
    }
    return Or.create(conditions);
  }
  /**
   * @param {Condition[]} conditions 
   * @returns {Condition}
   */
  static create(conditions) {
    if (conditions.length === 0) {
      return new _false.default();
    }
    if (conditions.length === 1) {
      return conditions[0];
    }
    return new Or(conditions);
  }
  toDcn() {
    return {
      call: ["or"],
      args: this.conditions.map(c => c.toDcn())
    };
  }
}
exports.default = Or;
