// @sap/ams/lib/cap/config/capAmsEnv.js
const path = require("node:path");
const os = require('node:os');
const fs = require('node:fs');

const cds = require("@sap/cds");

const config = require("./capAmsConfig");
const LocalLoader = require("../../../src/xcode/dcn-engine/build/commonjs/bundle/localloader").default;
const BundleLoader = require("../../../src/xcode/dcn-engine/build/commonjs/bundle/bundleloader").default;
const PolicyDecisionPoint = require("../../policyDecisionPoint");

const LOG = cds.log("ams");
const DEBUG = cds.debug("ams");

const ERROR_MSG_BUNDLE_LOADING = "Error loading AMS policy bundle: ";

let pdp;
/**
 * This function creates a pdp that is shared by the different components of the CAP plugin.
 * It is configured with a suitable BundleLoader based on the active CAP environment.
 * 
 * When mocked auth is active, it loads from the DCN root specified in the CAP environment.
 * When IAS auth is active, it loads from the corresponding AMS instance of the IAS credentials.
 * @returns {PolicyDecisionPoint} a PolicyDecisionPoint instance configured with a BundleLoader suitable for the authentication kind in the CAP environment
 */
function getPdp() {
	if (!pdp) {
		let bundleProvider;

		if (mockedAuthActive()) {
			const dcnRoot = getPathToDcn();
			if (!fs.existsSync(dcnRoot)) {
				fs.mkdirSync(dcnRoot, { recursive: true });
			}
			bundleProvider = new LocalLoader(dcnRoot, cds.watched);
		} else if (iasAuthActive()) {
			const credentials = cds.env.requires.auth.credentials;

			bundleProvider = new BundleLoader(
				new URL(`${credentials.url}/bundle-gateway/${credentials.authorization_instance_id}.dcn.tar.gz`),
				credentials.authorization_instance_id,
				credentials.certificate,
				credentials.key);
		}

		pdp = new PolicyDecisionPoint(null, null, bundleProvider);
		logBundleProviderEvents(bundleProvider);
	}

	return pdp;
}

function logBundleProviderEvents(bundleProvider) {
    bundleProvider.on("change", () => {
        DEBUG?.(`AMS policy bundle refreshed.`);
    });

    bundleProvider.on("requestFailed", async error => {
        const healthStatus = await pdp.getHealthStatus();
        if(healthStatus === PolicyDecisionPoint.OK || healthStatus === PolicyDecisionPoint.UPDATE_FAILED) {
            DEBUG?.(ERROR_MSG_BUNDLE_LOADING, error);
        } else {
            LOG.error(ERROR_MSG_BUNDLE_LOADING, error);
        }
    });
}

/** Returns the path to the local DCL files based on the CAP configuration or a default path if none was configured. */
function getPathToDcl() {
	const configuredDclFolder = process.env.AMS_DCL_ROOT ?? config.dclFolder;
	if (configuredDclFolder) {
		return configuredDclFolder;
	}

	// determine folder based on running locally or deployed
	if (cds.watched) {
		// running locally
		return path.join(cds.env?.build?.target === "." ? "gen" : cds.env?.build?.target, "srv", "ams", "dcl");
	} else {
		// running deployed
		return path.join(cds.root, "ams", "dcl");
	}
}

/** Returns the path to the local DCN files based on the CAP configuration or a default path if none was configured. */
function getPathToDcn() {
	return config.dcnRoot || path.resolve(os.homedir(), '.ams', 'dcn');
}

/** Heuristic to detect if application is running in production. */
function prodProfileActive() {
	return Array.isArray(cds.env.profiles) && cds.env.profiles.some(profile => profile.toLowerCase().includes("prod"));
}

/** Checks if active CAP profile is using mocked authentication. */
function mockedAuthActive() {
	return cds.env?.requires?.auth?.kind === "mocked";
}

/** Checks if active CAP profile is using IAS authentication. */
function iasAuthActive() {
	return cds.env?.requires?.auth?.kind === "ias" || cds.env?.requires?.auth?.kind === "ias-auth";
}

/** Checks if application is running with one of the supported authentication options required for AMS. */
function supportedAuthActive() {
	return mockedAuthActive() || iasAuthActive();
}

/** Connects to the AmsEventService if necessary and returns the service. */
let amsEventService;
async function getAmsEventService() {
	amsEventService ??= await cds.connect.to('AmsEventService');

	return amsEventService;
}


module.exports = {
	getPdp,
	getPathToDcl,
	prodProfileActive,
	mockedAuthActive,
	iasAuthActive,
	supportedAuthActive,
	getAmsEventService
};
