// @sap/ams/src/xcode/dcn-engine/build/commonjs/util/loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _promises = _interopRequireDefault(require("node:fs/promises"));
var _nodePath = _interopRequireDefault(require("node:path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var _default = exports.default = {
  async load(dcnDir) {
    const result = {
      policies: [],
      schemas: [],
      functions: [],
      data: {
        principal2policies: {}
      },
      manifest: {
        "revision": "0",
        "metadata": {
          "build_time": new Date().toISOString()
        }
      }
    };
    const dirContent = await _promises.default.opendir(dcnDir);
    for await (const dirEntry of dirContent) {
      const entryPath = _nodePath.default.join(dcnDir, dirEntry.name);
      if (dirEntry.isDirectory()) {
        const content = await this.load(entryPath);
        result.policies.push(...(content.policies ?? []));
        result.schemas.push(...(content.schemas ?? []));
        result.functions.push(...(content.functions ?? []));
      } else if (dirEntry.name === "data.json") {
        result.data = JSON.parse(await _promises.default.readFile(entryPath, "utf-8"));
      } else if (dirEntry.name.endsWith(".dcn")) {
        const content = JSON.parse(await _promises.default.readFile(entryPath, "utf-8"));
        result.policies.push(...(content.policies ?? []));
        result.schemas.push(...(content.schemas ?? []));
        result.functions.push(...(content.functions ?? []));
      }
    }
    return result;
  }
};
