// @sap/ams-dev/lib/cap/capAmsDevEnv.js
"use strict";
/* eslint-disable @typescript-eslint/no-var-requires */
const fs = require('fs');
const path = require('path');
const os = require('os');
const cds = require('@sap/cds');
const config = require('@sap/ams').cap.config;
const DEFAULT_OPA_PORT = 8181;
const POLICY_ASSIGNMENTS_FILENAME = 'policyAssignments';
class CapAmsDevEnv {
    get defaultOpaPort() {
        return DEFAULT_OPA_PORT;
    }
    get policyAssignmentFilename() {
        return POLICY_ASSIGNMENTS_FILENAME;
    }
    /**
     * Returns the path to which AMS compilation results are written during development given the build target folder of CAP.
     */
    getPathToAmsOutput() {
        let buildTarget = cds.env?.build?.target;
        if (buildTarget === '.') {
            // default CAP Java build target: build task compilations are placed "next" (.) to their source but we want the DCL files to be in a single, deterministic location
            buildTarget = path.join('srv', 'src', 'main', 'resources');
        }
        else if (cds.env?.['project-nature'] === 'nodejs') {
            // CAP Node.js build target: the express application itself will be compiled to <buildTarget>/srv and this is where the AMS folder must be located as well
            buildTarget = path.join(buildTarget, 'srv');
        }
        !buildTarget.endsWith('ams') && (buildTarget = path.join(buildTarget, 'ams'));
        return buildTarget;
    }
    /**
     * Returns the path to which DCL compilation results are written.
     * During runtime, use require('@sap/ams').cap.env.getPathToDcl() instead to access DCL files!
     */
    getPathToDclOutput() {
        return config.dclFolder ?? path.join(this.getPathToAmsOutput(), 'dcl');
    }
    /** Returns the path to custom DCL files. */
    get pathToCustomDcl() {
        return config.customDclFolder ?? path.join('ams', 'dcl');
    }
    /** Returns the path to the local OPA input files (REGO files, policy assignments file) based on the CAP configuration or a default path if none was configured. */
    get pathToLocalOpaInput() {
        return config.opa?.inputFolder ?? path.resolve(os.homedir(), '.opa');
    }
    /** Returns the path to the local REGO files based on the CAP configuration or a default path if none was configured. */
    get pathToLocalRego() {
        return path.join(this.pathToLocalOpaInput, 'rego');
    }
    /** Returns the path to the local policy assignment file based on the CAP configuration or a default path if none was configured. */
    get pathToLocalAssignments() {
        return path.join(this.pathToLocalOpaInput, 'policyAssignments');
    }
    /** Returns the port on which OPA should run based on the CAP configuration or the default port if none was configured. */
    get localOpaPort() {
        return config.opa?.port ?? this.defaultOpaPort;
    }
    /** Returns the address on which OPA should run based on the CAP configuration or the default port if none was configured. */
    get localOpaAddress() {
        return `http://127.0.0.1:${this.localOpaPort}`; // localhost does not work as it will be resolved using IPv6 in Node.Js >= 17
    }
    /**
     * Creates a write function that has the same behaviour as the one provided by CAP's BuildPlugin class.
     * @param targetFolder The target folder used to resolve paths based on the relative dest parameter of the 'to' function.
     * @param data The data that should be written. JS objects will be written as JSON.
     * @param dest The absolute or relative target path. Relative paths will be resolved from targetFolder.
     */
    createWriteFunction(targetFolder) {
        return (data) => ({
            to: async (dest) => {
                if (!path.isAbsolute(dest)) {
                    dest = path.resolve(targetFolder, dest);
                }
                if (!fs.existsSync(path.dirname(dest))) {
                    fs.mkdirSync(path.dirname(dest), { recursive: true });
                }
                return fs.promises.writeFile(dest, typeof data === 'object' && !Buffer.isBuffer(data) ? JSON.stringify(data, null, 2) : data);
            }
        });
    }
    /**
     * Creates a copy function that has the same behaviour as the one provided by CAP's BuildPlugin class.
     * @param sourceFolder The source folder used to resolve paths based on the relative src parameter.
     * @param targetFolder The target folder used to resolve paths based on the relative dest parameter of the 'to' function.
     * @param src The absolute source path of the file or directory to copy. Relative paths will be resolved from sourceFolder.
     * @param dest The absolute or relative target path. Relative paths will be resolved from targetFolder.
     */
    createCopyFunction(sourceFolder, targetFolder) {
        return (src) => ({
            to: async (dest) => {
                if (!path.isAbsolute(src)) {
                    src = path.resolve(sourceFolder, src);
                }
                if (!path.isAbsolute(dest)) {
                    dest = path.resolve(targetFolder, dest);
                }
                return fs.promises.cp(src, dest, { recursive: true });
            }
        });
    }
}
module.exports = new CapAmsDevEnv();
