// @sap/ams/lib/roles/RolesProvider.js
const Attributes = require("../attributes");
const AttributeName = require("../attributeName");
const Call = require("../call");

const LOG = require("../logger");

const RolesCache = require("./RolesCache");
const PolicyDecisionPoint = require("../policyDecisionPoint");

class RolesProvider {
    #pdp;
    #cache;

    constructor(pdp = new PolicyDecisionPoint()) {
        this.#pdp = pdp;        
    }

    withRolesCache(cache = new RolesCache()) {
        this.#cache = cache;
        return this;
    }

    async getRoles(principle) {
        let roles;
        
        if(this.#cache) {
            const cacheKey = principle.toCacheKey();
            roles = this.#cache.get(cacheKey);
    
            if (roles !== null) {                
                LOG.debug("Using cached AMS user roles.");
                return roles;
            } else {
                roles = await this.computeRoles(principle);
                this.#cache.set(cacheKey, roles);
            }
        } else {
            roles = await this.computeRoles(principle);
        }

        return roles;
    }

    async computeRoles(principle) {
        LOG.debug(`Computing AMS user roles for ${principle.userId} on tenant ${principle.tenant}.`);

        const attrs = new Attributes()
            .setResource("$SCOPES")
            .setUnknowns([["$dcl", "action"], ["$app"], ["$env"]])
            .setTenant(principle.tenant)
            .setPrincipalToPolicies([principle.tenant, principle.userId]);

        let answer;
        try {
            answer = await this.#pdp.allowFilterClause(attrs);
        } catch (e) {
            throw new Error("Error during policy evaluation.", { cause: e });
        }

        if (typeof answer.condition !== "object") {
            if (answer.condition === false) {
                LOG.debug("DCN Engine returned unconditionally false when queried for AMS user roles.", answer);
                return [];
            }

            throw new Error("DCN Engine returned an unknown response format.");
        }

        LOG.debug("DCN Engine returned this answer when queried for AMS user roles.", answer);
        return extractScopes(answer.condition);
    }

}

function extractScopes(condition) {
    let scopes;

    const call = Call.fromCondition(condition);        
    if (call.getType() === Call.types.EQ) {
        // result contains single scope
        scopes = [extractScope(call)];
    } else if (call.getType() === Call.types.OR) {
        // result contains list of scopes
        scopes = call.getArguments().map(c => extractScope(c));
    }

    return scopes.filter(scope => scope !== null);
}

function extractScope(call) {
    const [arg0, arg1] = call.getArguments();

    if (AttributeName.isAttributeName(arg0) && arg0.toString() === "$dcl.action") {
        return arg1;
    } else if (AttributeName.isAttributeName(arg1) && arg1.toString() === "$dcl.action") {
        return arg0;
    }

    LOG.warn("AMS scopes condition returned by DCN Engine contained unparseable call. Please make sure to only grant scopes to $SCOPES in DCL without (WHERE) condition!", call);
    return null;
}

module.exports = RolesProvider;
