// @sap/ams/lib/auditlog/index.js
"use strict";

const pdpEvents = require("../policyEvents").eventName;
const helpers = require("./utils");

/** optional conf object:
 * {
 * 	accessChannel: "web",
 *  type: "AmsPolicyEvaluation",
 *  tenant: "tenant_id",
 * 	byUser: "user_id",
 *  dataSubject: {
 *		type: "application-oauth2-client",
 *		id: { client_id: "1287z1827z3" }
 *	},
 *  callTransformer: {} => (...)
 * }
 */

/**
 * This class takes a Policy Decision Point instance in the constructor to
 * auditlogs every allow + allowFilterClause request by registering on those events
 */
class PdpEventWriter {
	/**
	 * @param {Object} pdp Policy Decision Point instance to retrieve events
 	 * @param {Object} auditLog object created by audit log service
	 * @param {Function} callback containing the error as parameter
	 * @param {Object} conf optional config conf.tenant and conf.byUser are necessary for auditlog standard plan
	 */
	constructor(pdpInstance, auditLog, callback, conf = {}) {
		this.pdp = pdpInstance;
		this._eventFunc = (pdpRes) => {
			writeMessage(auditLog, pdpRes, callback, conf);
		};
		this.pdp.on(pdpEvents.POLICY_EVALUATION, this._eventFunc);
	}

	// unregister the listener of the pdp
	unregister() {
		this.pdp.removeListener(pdpEvents.POLICY_EVALUATION, this._eventFunc);
	}
}

/**
 * Writes auditlog with given pdp result object and optional other conf
 * @param {Object} auditLog
 * @param {Object} pdpRes
 * @param {Function} callback containing the error as parameter
 * @param {Object} conf optional additional config
 */
function writeMessage(auditLog, pdpRes, callback, conf = {}) {
	try {
		const { idObject, logAttributeName } = helpers.formatPdpResponse(pdpRes, conf);
		const dataSubject = helpers.getDataSubject(pdpRes.tokenClientId, conf);
		const tenant = conf.tenant ? conf.tenant : "$PROVIDER"; // oauth plan uses $PROVIDER
		const user = conf.byUser ? conf.byUser : "$USER"; // oauth plan uses $USER

		auditLog.read({
			type: conf.type ? conf.type : "AmsPolicyEvaluation",
			id: idObject
		}).attribute({ name: logAttributeName, successful: pdpRes.isSucceeded })
			.dataSubject(dataSubject)
			.accessChannel(conf.accessChannel ? conf.accessChannel : "web")
			.tenant(tenant)
			.by(user)
			.log(err => {
				if (err) callback(err);
			});
	} catch(err) {
		callback(err);
	}
}

module.exports = { PdpEventWriter, writeMessage };


/**
 * Auditlog example
 * https://github.wdf.sap.corp/CPSecurity/Knowledge-Base/blob/master/13_AuthorizationManagementService(AMS)/AuditLogging.md
 * {
 *   "uuid": "5e05c864-1aa2-430d-ab73-5f1da0d0346f",
 *   "user": "d9403e85-2029-46f1-9c09-ee32e881c081",
 *   "time": "2021-04-14T12:28:32.041Z",
 *   "channel": "web",
 *   "object": {
 *      "type": "AmsPolicyEvaluation",
 *      "id": {
 *          "operation": "allow",
 *          "kind": "QUERY",
 *          "ignoring": "[]",
 *          "unknowns": "[]",
 *          "$dcl.resource": "salesOrders",
 *          "$dcl.action": "read",
 *          "$dcl.principal2policies": "[4b0c2b7a-1279-4352-a68d-a9a228a4f1e9, d9403e85-2029-46f1-9c09-ee32e881c081]",
 *          "$dcl.tenant": "4b0c2b7a-1279-4352-a68d-a9a228a4f1e9",
 *          "access": "grant",
 *          "accessResult": "true"
 *      }
 *  },
 *  "data_subject": {
 *      "type": "application-oauth2-client",
 *      "id": {
 *          "client_id": "28c17dfe-dc96-47fe-83ad-54f52d4bbd57"
 *      }
 *  },
 *  "attributes": [
 *      {
 *          "name": "access",
 *          "successful": true
 *      }
 *  ],
 *  "attachments": [],
 *  "tenant": "4b0c2b7a-1279-4352-a68d-a9a228a4f1e9"
 * }
*/
