// @sap/ams/lib/auditlog/utils.js
"use strict";

const pdpEvents = require("../policyEvents").eventName;
const Call = require("../call");
const isAttributeName = require("../attributeName").isAttributeName;

// returns auditlog object.id and attributes.name
module.exports.formatPdpResponse = function(pdpRes, conf = {}) {
	let idObject = { operation: pdpRes.operation };
	let logAttributeName = "unknown";

	switch (pdpRes.operation) {
	case pdpEvents.ALLOW:
		logAttributeName = "access";
		idObject.kind = "QUERY";
		idObject.access = pdpRes.isSucceeded ? "grant" : "error";
		idObject.accessResult = pdpRes.result?.toString();
		break;
	case pdpEvents.ALLOW_PARTIAL:
		logAttributeName = "conditionalAccess";
		idObject.kind = "Filter";
		idObject.access = pdpRes.isSucceeded ? "constraint" : "error";
		if (typeof(pdpRes.result) !== "boolean") {
			idObject.accessResult = transformCondition(pdpRes.result, conf.callTransformer);
		} else {
			idObject.accessResult = pdpRes.result?.toString();
		}
		break;
	default:
		idObject.access = pdpRes.isSucceeded ? "true" : "error";
		idObject.accessResult = pdpRes.result;
	}

	const cleanedAttr = JsonValuesToString(flattenJSON(pdpRes.attributesJSON));
	idObject = Object.assign(idObject, cleanedAttr);
	return { idObject, logAttributeName };
};

module.exports.getDataSubject = function(attrClientId, conf) {
	if (conf.dataSubject) {
		return conf.dataSubject;
	}
	// if attribute.setTokenInfo() was called
	if (attrClientId) {
		return { type: "application-oauth2-client", id: { client_id: attrClientId } };
	}
	return { type: "unknown", id: { client_id: "unknown" } };
};

function JsonValuesToString(obj = {}, res = {}) {
	for(const key in obj) {
		res[key] = obj[key]?.toString();
	}
	return res;
}

function flattenJSON(obj = {}, res = {}, extraKey = "") {
	for(const key in obj) {
		if (typeof obj[key] !== "object") {
			res[extraKey + key] = obj[key];
		} else {
			flattenJSON(obj[key], res, `${extraKey}${key}.`);
		}
	}
	return res;
}

function transformCondition(condition, customTransformer) {
	const call = Call.fromCondition(condition);
	return Call.transform(call, customTransformer ? customTransformer : callToString);
}

// e.g. and(and(eq($app.COMPANY_CODE, 0001), eq($app.PURCHASE_ORG, SAPSE)), eq($app.PRODUCT_CATEGORY, IT Product))
function callToString(item) {
	if (Call.isCall(item)) {
		const callChildren = [item.getType(), "("];
		for (let i = 0; i < item.getArgumentCount(); i++) {
			callChildren.push(item.getArgument(i));
			if (i !== item.getArgumentCount() -1) {
				callChildren.push(", ");
			}
		}
		callChildren.push(")");
		return callChildren;
	}
	else if (isAttributeName(item)) {
		return item.toString();
	}
}
