// @sap/ams-dev/lib/stopOpa.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRunningOpasByConfig = exports.stopOpaByConfig = exports.stopOpa = void 0;
const yargs_1 = __importDefault(require("yargs"));
const util = __importStar(require("./util"));
let logger = util.getLogger('error');
async function stopOpa(args) {
    const options = (0, yargs_1.default)(args)
        .usage('Usage: stop-opa -p [PORT] -n [NAME]')
        .options({
        p: { alias: 'port', describe: 'stops the opa process running on this port', type: 'string', demandOption: false },
        n: { alias: 'name', describe: 'stops the opa process with this given name', type: 'string', demandOption: false },
        l: { alias: 'log-level', describe: 'log level', type: 'string', default: 'error', demandOption: false, choices: ['debug', 'info', 'warn', 'error', 'silent'] },
        d: { alias: 'dry-run', describe: 'list running instances to delete', type: 'boolean', default: false, demandOption: false }
    })
        .example('stop-opa -p 8181', 'stopps opa on port 8181')
        .parseSync();
    logger = util.getLogger(options.l);
    const config = {
        name: options.n,
    };
    if (options.p) {
        config.port = parseInt(options.p);
    }
    if (options.d) {
        const runningOpas = getRunningOpasByConfig(config);
        console.log(JSON.stringify(runningOpas));
        return;
    }
    return stopOpaByConfig(config);
}
exports.stopOpa = stopOpa;
async function stopOpaByConfig(options) {
    let allPrcs = util.opaProcesses();
    if (allPrcs.length == 0) {
        logger.info('there are no opa instances running in background');
        return [];
    }
    const prcsToKill = getRunningOpasByConfig(options);
    prcsToKill.forEach(killPrc => {
        killProcess(killPrc.pid, killPrc.port, killPrc.name);
        allPrcs = allPrcs.filter(prc => prc.pid != killPrc.pid);
    });
    if (prcsToKill.length == 0) {
        logger.info('found no process to stop');
    }
    else {
        logger.info(`stopped ${prcsToKill.length} process(es)`);
    }
    util.persistProcesses(allPrcs);
}
exports.stopOpaByConfig = stopOpaByConfig;
function killProcess(pid, port, name) {
    try {
        process.kill(pid);
    }
    catch (err) {
        logger.error('killing opa instance was unsuccessful');
        logger.error(`pid: ${pid}, port: ${port}, name: ${name}`);
        logger.error(err);
    }
}
function getRunningOpasByConfig(opts) {
    const opaPrcs = util.opaProcesses();
    return opaPrcs.filter(prc => (opts.port && parseInt(prc.port) == opts.port) ||
        (opts.name && prc.name == opts.name) ||
        (!opts.name && !opts.port));
}
exports.getRunningOpasByConfig = getRunningOpasByConfig;
