// @sap/ams-dev/lib/updateBins.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const stream = __importStar(require("stream"));
const zlib = __importStar(require("zlib"));
const tar = __importStar(require("tar-fs"));
const util_1 = require("util");
const finished = (0, util_1.promisify)(stream.pipeline);
async function start() {
    const pkg = JSON.parse(fs.readFileSync('package.json', { encoding: 'utf8', flag: 'r' }));
    const opaVer = pkg.config.opaVersion;
    const compilerVer = pkg.config.dclCompilerVersion;
    const opaVerPath = path.join('resources', 'opa', 'VERSION');
    const compilerVerPath = path.join('resources', 'dcl-compiler', 'VERSION');
    // create dirs if not exist
    if (!fs.existsSync(path.dirname(opaVerPath))) {
        fs.mkdirSync(path.dirname(opaVerPath), { recursive: true });
    }
    if (!fs.existsSync(path.dirname(compilerVerPath))) {
        fs.mkdirSync(path.dirname(compilerVerPath), { recursive: true });
    }
    await Promise.all([
        updateBinary('dcl-compiler', compilerVerPath, compilerVer),
        updateBinary('opa', opaVerPath, opaVer)
    ]);
}
async function updateBinary(name, versionPath, desiredVersion) {
    if (fs.existsSync(versionPath)) {
        const currentVersion = fs.readFileSync(versionPath, { encoding: 'utf8', flag: 'r' });
        if (currentVersion === desiredVersion) {
            console.log(`${name} is up to date`);
            return;
        }
        else {
            console.log(`${name} is outdated: current: ${currentVersion} desired: ${desiredVersion}`);
        }
    }
    rmDirContent(path.join('resources', name));
    console.log(`downloading new ${name} version ${desiredVersion}`);
    switch (name) {
        case 'opa':
            await downloadOpa(desiredVersion);
            break;
        case 'dcl-compiler':
            await downloadDclCompiler(desiredVersion);
            break;
    }
    fs.writeFileSync(path.join('resources', name, 'VERSION'), desiredVersion);
}
async function downloadOpa(version) {
    const downloadOpa = async (os) => {
        const osPath = path.join('resources', 'opa', os);
        fs.mkdirSync(osPath);
        const OPA_DOWN_URL = `https://int.repositories.cloud.sap/artifactory/build-snapshots/com/sap/golang/github/wdf/sap/corp/cpsecurity/cas-opa-sap/${version}/cas-opa-sap-${version}-${os}-amd64.tar.gz`;
        const tarGzPath = path.join('resources', 'opa', os, 'opa.tar.gz');
        await downloadFile(OPA_DOWN_URL, tarGzPath);
        extractTarGzFile(tarGzPath, path.resolve(osPath));
    };
    await Promise.all([
        downloadOpa('darwin'),
        downloadOpa('linux'),
        downloadOpa('windows'),
    ]);
}
async function downloadDclCompiler(version) {
    const COMPILER_DOWN_URL = `https://int.repositories.cloud.sap/artifactory/build-snapshots/com/sap/cloud/security/ams/dcl/compiler/${version}/compiler-${version}-full.jar`;
    await downloadFile(COMPILER_DOWN_URL, path.join('resources', 'dcl-compiler', 'dcl.jar'));
}
function rmDirContent(directory) {
    if (fs.existsSync(directory))
        fs.rmSync(directory, { recursive: true, force: true });
    fs.mkdirSync(directory);
}
async function downloadFile(url, dest) {
    try {
        const request = await axios_1.default.get(url, {
            responseType: 'stream',
        });
        await finished(request.data, fs.createWriteStream(dest));
    }
    catch (error) {
        throw (`download of ${url} failed ${error}`);
    }
}
function extractTarGzFile(path, dest) {
    const fileContents = fs.createReadStream(path);
    const unzip = zlib.createGunzip();
    const extract = fileContents.pipe(unzip).pipe(tar.extract(dest, { strip: 1 }));
    extract.on('finish', function () {
        fs.unlinkSync(path);
    });
}
start();
