// @sap/ams/lib/attributeKeys.js
/**
 * {
 *   "$dcl": {
 *     "action":                "read",
 *     "resource":              "SalesOrder",
 *     "tenant":                "12345",
 *     "scopeFilter":           ["ams.readAllSalesOrders"]"
 *     "principal2policies":    ["zone_a","user_a"],
 *     "policies":              ["ams.readAllSalesOrders"]
 *   },
 *   "$app": {
 *     country": "DE"
 *   },
 *   "$env": {
 *     "$user": {
 *       "user_uuid": "123",
 *       "email": "alice@sap.com",
 *       "groups": ["g1", "g2"]
 *     }
 *   }
 *   Only used in pdp.allowFilterClause
 *   "unknowns": [
 *      "input[\"$dcl\"]"
 *   ],
 *   "ignores": [
 *       "input[\"$dcl\"][\"resource\"]"
 *   ]
 * }
 **/

/**
 * DCL section
 */
const DCL_SECTION_KEY = "$dcl";
const DCL_ACTION_KEY = "action"; // A string is expected as value
const DCL_RESOURCE_KEY = "resource"; // A string is expected as value
const DCL_TENANT_KEY = "tenant"; // A string is expected as value
const DCL_POLICIES_KEY = "policies"; // An array of strings is expected as value
const DCL_PRINCIPAL_TO_POLICIES_KEY = "principal2policies"; // An array of strings is expected as value
const DCL_SCOPE_FILTER_KEY = "scopeFilter"; // An array of strings is expected as value
 
/**
 * APP section
 */
const APP_SECTION_KEY = "$app";
 
/**
 * ENV section
 */
const ENV_SECTION_KEY = "$env";
const ENV_USER_KEY = "$user";
const ENV_USER_USER_UUID_KEY = "user_uuid";
const ENV_USER_EMAIL_KEY = "email";
const ENV_USER_GROUPS_KEY = "groups";
 
/**
 * UNKNOWNS and IGNORES section
 */
const UNKNOWNS_SECTION_KEY = "unknowns";
const IGNORES_SECTION_KEY = "ignores";

const keys = {
	DCL_SECTION: DCL_SECTION_KEY,
	DCL_ACTION: DCL_ACTION_KEY,
	DCL_RESOURCE: DCL_RESOURCE_KEY,
	DCL_TENANT: DCL_TENANT_KEY,
	DCL_POLICIES: DCL_POLICIES_KEY,
	DCL_PRINCIPAL_TO_POLICIES: DCL_PRINCIPAL_TO_POLICIES_KEY,
	DCL_SCOPE_FILTER: DCL_SCOPE_FILTER_KEY,
	APP_SECTION: APP_SECTION_KEY,
	ENV_SECTION: ENV_SECTION_KEY,
	ENV_USER: ENV_USER_KEY,
	ENV_USER_USER_UUID: ENV_USER_USER_UUID_KEY,
	ENV_USER_EMAIL: ENV_USER_EMAIL_KEY,
	ENV_USER_GROUPS: ENV_USER_GROUPS_KEY,
	UNKNOWNS_SECTION: UNKNOWNS_SECTION_KEY,
	IGNORES_SECTION: IGNORES_SECTION_KEY
};

module.exports = keys;
