// @sap/ams/lib/dclPackageName.js
"use strict";

const ErrType = require("./errors/errorTypes");

const DCL_PACKAGE_SEPARATOR_CHAR = ".";
const DCL_PACKAGE_SEPARATOR_CHAR_FOR_URL = "/";

/**
 * @param {String} name dcl package name
 * @returns the dcl package name with dots replaced by "/"
 */
function toUrl(name) {
	const arr = name.split(DCL_PACKAGE_SEPARATOR_CHAR);
	return arr.join(DCL_PACKAGE_SEPARATOR_CHAR_FOR_URL);
}

/**
 * @throws {DCLPackageNameError}
 * @param {String} name dcl package name
 * calls reject function with error if the dcl package name contains illegal characters
 */
function isQualifiedPackage(name) {
	for (const identifier of name.split(DCL_PACKAGE_SEPARATOR_CHAR))
		isIdentifier(identifier);
}

/**
 * @throws {DCLPackageNameError}
 * @param {String} name package name
 */
function isIdentifier(name) {
	if (!name) {
		throw new ErrType.DCLPackageNameError("DCL package identifier name may not be empty");
	}
	if (!isIdentifierStart(name.charAt(0))) {
		throw new ErrType.DCLPackageNameError(`DCL package identifier:${name} starts with illegal character`);
	}

	for(let i=1;i<name.length;++i) {
		if(!isIdentifierPart(name.charAt(i))) {
			throw new ErrType.DCLPackageNameError(`DCL package:${name} contains illegal character`);
		}
	}
}

function isIdentifierStart(ch) {
	if (ch == "_") {
		return true;
	}
	return isLowercaseLetter(ch);
}

function isIdentifierPart(ch) {
	if (ch == "_") {
		return true;
	}
	return isLowercaseLetterOrNumber(ch);
}

function isLowercaseLetter(ch) { 
	return /^[a-z]$/.test(ch);
}

function isLowercaseLetterOrNumber(ch) { 
	return /^[a-z0-9]$/.test(ch);
}

module.exports = {
	toUrl,
	isQualifiedPackage
};
