// @sap/ams/lib/policyEvents.js
"use strict";

const eventName = {
	ANY:                "pdpEvaluation", // emits every pdp event
	BEFORE_EVALUATION:	"beforeEvaluation",
	POLICY_EVALUATION:  "amsPolicyEvaluation", // emits allow and allowPartial events
	ALLOW:              "allow",
	ALLOW_PARTIAL:      "allowPartial",
	PING:               "ping",
	GET_VERSION:        "getVersion",
	GET_STATUS:         "getStatus"
};

const EVENT_POLICY_EVALUATION_GROUP = [
	eventName.ALLOW,
	eventName.ALLOW_PARTIAL
];

/**
 * Emits an event of the pdp, automatically also emits on multiple eventNames
 * @param {EventEmitter Object} eventEmitter 
 * @param {evenObject} eventName 
 */
function emitEvent(eventEmitter, eventObject) {
	// emit the normal event
	eventEmitter.emit(eventObject.operation, eventObject);
	// emit the event on pdpEvaluation
	eventEmitter.emit(eventName.ANY, eventObject);
	if (EVENT_POLICY_EVALUATION_GROUP.indexOf(eventObject.operation) > -1) {
		eventEmitter.emit(eventName.POLICY_EVALUATION, eventObject);
	}
}

/**
 * @param {String} operation operation string, e.g.: "allow"
 * @param {Object} pdpResult result of the pdp (opa) response
 * @param {JSON} httpBody the json which was sent to the opa
 * @param {Attributes} attr the attributes which were sent to the opa
 * @param {String} dclpkg in allow or allowFilterClause specified DCL package
 * @param {bool} successful has an error been thrown in the request or not
 * @param {String} pdpURL URL of the server running the policy decision point
 * @returns {JSON} eventObject for logging
 */
function createEventObject(pdpURL, operation, dclpkg = null, httpBody = null, attr = null, successful = null, pdpResult = {}) {
	return {
		"operation": operation, // Name of the event
		"result": pdpResult, // response from pdp
		"httpBody": httpBody, // body sent to the pdp
		"attributesJSON": attr ? attr.getJSON() : {}, // attributes sent to the pdp if any
		"dclPackage": dclpkg, // set DCL package
		"isSucceeded": successful, // boolean - false if an error has occurred
		"tokenClientId": (attr && attr.getTokenInfo()) ? attr.getTokenInfo().getClientId() : null,
		"pdpURL": pdpURL // set pdp URL
	};
}


module.exports = {
	eventName,
	emitEvent,
	createEventObject
};
